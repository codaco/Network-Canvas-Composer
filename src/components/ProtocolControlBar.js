import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Icon, Spinner } from '@codaco/ui';
import { getProtocol } from '@selectors/protocol';
import { actionCreators as protocolsActions, actionLocks as protocolsLocks } from '@modules/protocols';
import { selectors as statusSelectors } from '@modules/ui/status';
import ControlBar from './ControlBar';

const RightArrow = <Icon name="arrow-right" />;

const ProtocolControlBar = ({
  saveProtocol,
  hasUnsavedChanges,
  hasAnyStages,
  show,
  isSaving,
}) => {
  const saveProps = !isSaving &&
    {
      icon: RightArrow,
      iconPosition: 'right',
    };
  return (
    <ControlBar
      show={show && hasUnsavedChanges}
      buttons={[
        <Button
          onClick={saveProtocol}
          color="white"
          data-variant="save"
          disabled={!hasAnyStages || isSaving}
          {...saveProps}
        >
          {!isSaving && 'Save'}
          {isSaving && <Spinner /> }
        </Button>,
      ]}
    />
  );
};

ProtocolControlBar.propTypes = {
  saveProtocol: PropTypes.func.isRequired,
  hasUnsavedChanges: PropTypes.bool.isRequired,
  hasAnyStages: PropTypes.bool.isRequired,
  isSaving: PropTypes.bool.isRequired,
  show: PropTypes.bool,
};

ProtocolControlBar.defaultProps = {
  show: true,
};

const mapStateToProps = state => ({
  hasUnsavedChanges: (
    state.session.lastChanged >
    state.session.lastSaved
  ),
  hasAnyStages: getProtocol(state).stages.length > 0,
  isSaving: statusSelectors.getIsBusy(state, protocolsLocks.saving),
});

const mapDispatchToProps = {
  saveProtocol: protocolsActions.saveAndBundleProtocol,
};

export { ProtocolControlBar };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProtocolControlBar);
