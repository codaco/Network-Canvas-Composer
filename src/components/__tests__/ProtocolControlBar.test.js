/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { ProtocolControlBar } from '../ProtocolControlBar';

const mockProps = {
  saveProtocol: () => {},
  isSaving: false,
  isDisabled: false,
  showControlBar: true,
};

describe('<ProtocolControlBar />', () => {
  it('can render', () => {
    const component = shallow(<ProtocolControlBar {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
