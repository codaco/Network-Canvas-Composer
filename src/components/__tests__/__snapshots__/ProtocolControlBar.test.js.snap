// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProtocolControlBar /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProtocolControlBar
    hasAnyStages={false}
    hasUnsavedChanges={false}
    saveProtocol={[Function]}
    show={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "buttons": Array [
        <Button
          color="white"
          content=""
          data-variant="save"
          icon={
            <Icon
              className=""
              color=""
              name="arrow-right"
              style={Object {}}
            />
          }
          iconPosition="right"
          onClick={[Function]}
          size=""
          type="button"
        >
          <ForwardRef(MotionComponent)
            animate="ready"
            initial="ready"
          >
            <ForwardRef(MotionComponent)
              variants={
                Object {
                  "busy": Object {
                    "opacity": 0,
                  },
                  "ready": Object {
                    "opacity": 1,
                  },
                }
              }
            >
              Save
            </ForwardRef(MotionComponent)>
            <ForwardRef(MotionComponent)
              className="button__busy"
              variants={
                Object {
                  "busy": Object {
                    "opacity": 1,
                  },
                  "ready": Object {
                    "opacity": 0,
                  },
                }
              }
            >
              <Spinner
                large={false}
                small={false}
              />
            </ForwardRef(MotionComponent)>
          </ForwardRef(MotionComponent)>
        </Button>,
      ],
      "className": "",
      "flip": false,
      "secondaryButtons": null,
      "show": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "buttons": Array [
          <Button
            color="white"
            content=""
            data-variant="save"
            icon={
              <Icon
                className=""
                color=""
                name="arrow-right"
                style={Object {}}
              />
            }
            iconPosition="right"
            onClick={[Function]}
            size=""
            type="button"
          >
            <ForwardRef(MotionComponent)
              animate="ready"
              initial="ready"
            >
              <ForwardRef(MotionComponent)
                variants={
                  Object {
                    "busy": Object {
                      "opacity": 0,
                    },
                    "ready": Object {
                      "opacity": 1,
                    },
                  }
                }
              >
                Save
              </ForwardRef(MotionComponent)>
              <ForwardRef(MotionComponent)
                className="button__busy"
                variants={
                  Object {
                    "busy": Object {
                      "opacity": 1,
                    },
                    "ready": Object {
                      "opacity": 0,
                    },
                  }
                }
              >
                <Spinner
                  large={false}
                  small={false}
                />
              </ForwardRef(MotionComponent)>
            </ForwardRef(MotionComponent)>
          </Button>,
        ],
        "className": "",
        "flip": false,
        "secondaryButtons": null,
        "show": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
